torch/distributed/_shard/checkpoint/filesystem.py

```diff
@@ -0,0 +1,131 @@
+import os
+import operator
+import pickle
+from typing import List, Optional, cast
+
+import torch
+from torch import Tensor
+import torch.distributed as dist
+from torch.futures import Future
+from pathlib import Path
+
+from .metadata import (
+    BytesReadRequest,
+    BytesWriteRequest,
+    Metadata,
+    TensorReadRequest,
+    TensorWriteRequest,
+)
+from .storage import StorageReader, StorageWriter
+
+def _get_file_storage_path(
+    storage_path: str,
+    storage_key: str
+) -> Path:
+    return (Path(storage_path) / storage_key)
+
+class FileSystemWriter(StorageWriter):
```

Comment:
This example seems a bit flimsy to certain extent. Even though it's just a prototype, we probably need something better than a toy:

1. It does not cover failure handling logic. Just a quick example: what's the handling logic if `torch.save()` fails? How does caller handle that?
2. This does not show the capability of `write_metadata()` and arguably it should - for example the file system writer should probably have an example to show how to handle Windows case-insensitive file system when model wants to save `foo` and `Foo` (how to prevent one from overwriting the other using storage metadata).
3. The reader/writer does not say "I'm an SPMD implementation" (e.g. injecting `pg` in constructor) yet they just assume that which may make it hard to support generic resharding impl (like async training using RPC).

Reply:
1. Failure handing still under design and I plan to work on it next. Tracking it here: https://github.com/pytorch/pytorch/issues/76395

2. The current design uses generated unique storage_keys (it hashes keys) so there's no need to handle case-insensitive filesystems and I believe this is the way to go as it solves other issues such as invalid characters and FQNs that are relative paths (IE `../foo`).

3. If Reader / Writer needs to use collectives, they must assume they are operating over the root PG. The current setup supports resharding ShardedTensors just fine. I'm not sure what you mean by generic resharding in this context.

As we introduce collectives to load/save state_dict in order to coordinate saving we'll have to allow a PG to be passed in.

Reply:
> The current design uses generated unique storage_keys (it hashes keys) so there's no need to handle case-insensitive filesystems and I believe this is the way to go as it solves other issues such as invalid characters and FQNs that are relative paths (IE ../foo).

So this is not how we internally handle storage metadata today and not having a plan to support that would be a hard blocker for adoption (in a sense that no production workload can support this OSS solution until there's a story).

Today say a param `foo` is sharded into 2 ranks with shard FQN `foo.1` and `foo.2`, the storage may further chunk each shard to 2 files, or store the 2 shards into 1 file for optimization (if the shards are small and stacked into the same rank somehow).

Let's say it's the former so effectively we would store:

```
{
  "foo": {
    "dtype": "int32",
    "shards": [
      "foo.1": {
        "sizes": [5],
        "chunks": [
          {
            "filename": "foo.1_1",
            "offset": 0,
            "length": 10,
          },
          {
            "filename": "foo.1_2",
            "offset": 0,
            "length": 10,
          }
        ]
      },
      "foo.2": {
        "sizes": [5],
        "chunks": [
          {
            "filename": "foo.2_1",
            "offset": 0,
            "length": 20,
          },
        ]
      }
    ]
  }
}
```

Could you explain how this case can be supported properly when it comes to metadata transformation?

3. If Reader / Writer needs to use collectives, they must assume they are operating over the root PG. The current setup supports resharding ShardedTensors just fine. I'm not sure what you mean by generic resharding in this context.

In this implementation, it's not clear what the reader/writer responsibility is (which is what @simpkins points out above). Clearly in the example impl, the `write_tensor()` part is called in every rank, while the `write_metadata()` part is not clear at all.

In the api doc, it says

```
    3) write_tensors.
    4) Wait for (2) and (3) futures. If either fail, abort checkpoint.
    5) (if rank == 0) write_metadata().
```
which indicates **the caller** is responsible for checking if this needs to be called in specific ranks, yet in the example, it's the implementation.

This just shows that the proposal itself is not clear about who is responsible for that.

By generic, what I meant was the implementation should not be bound to process group as that seems to assume the synchronization is done by collective only. This is not universally true. PyTorch itself doesn't say all training needs to be full-sync using SPMD. If synchronization is done via RPC, this assumption would break. In theory, the same storage writer should handle both cases with zero code change, so it's better for caller to ensure the rank 0 check + additional synchronization if any across hosts/ranks.

BTW, can we not resolve comments until there's a real resolution?

Reply:
Oh, and I'm moving the `rank 0` check away from FileSystemWriter